src/gridcoin/voting/registry.cpp

```diff
@@ -303,12 +309,270 @@ int64_t PollReference::Age(const int64_t now) const
+                }
+
+                scaled_network_magnitude = superblock.m_cpids.TotalScaledMagnitude();
+            }
+        }
+
+        // Please refer to https://gridcoin.us/assets/docs/grc-bluepaper-section-1.pdf equations 1 and 16 and footnote 5.
+        // This method of computing net_weight is from first principles using the target from the nBits representation
+        // recorded in the index, rather than the GetEstimatedNetworkWeight() function, which uses double fp arithmetic.
+        arith_uint256 target;
+        target.SetCompact(pindex->nBits);
+
+        arith_uint256 net_weight = ~arith_uint256() / arith_uint256(450) / target * arith_uint256(COIN);
+
+        arith_uint256 money_supply = pindex->nMoneySupply;
+
+        tally_active_vote_weight(net_weight, money_supply, scaled_pool_magnitude, scaled_network_magnitude);
+
+        // If voting logging category is active, log the first block and every superblock
+        if (blocks == 1 || pindex->IsSuperblock()) {
+            LogPrint(BCLog::LogFlags::VOTE, "INFO: %s: tally_active_vote_weight: net_weight = %f, money_supply = %f, "
+                                            "pool_magnitude = %f, network_magnitude = %f, block height = %i, "
+                                            "blocks = %u, active_vote_weight_tally = %f, active_vote_weight = %f.",
+                     __func__,
+                     net_weight.getdouble() / (double) COIN,
+                     money_supply.getdouble() / (double) COIN,
+                     scaled_pool_magnitude.getdouble() / 100.0,
+                     scaled_network_magnitude.getdouble() / 100.0,
+                     pindex->nHeight,
+                     blocks,
+                     active_vote_weight_tally.getdouble() / (double) COIN,
+                     (active_vote_weight_tally / arith_uint256(blocks)).getdouble() / (double) COIN
+                     );
+        }
+
+        // Log the last block and break if at pindex_end.
+        if (pindex == pindex_end) {
+            LogPrint(BCLog::LogFlags::VOTE, "INFO: %s: tally_active_vote_weight: net_weight = %f, money_supply = %f, "
+                                            "pool_magnitude = %f, network_magnitude = %f, block height = %i, "
+                                            "blocks = %u, active_vote_weight_tally = %f, active_vote_weight = %f.",
+                     __func__,
+                     net_weight.getdouble() / (double) COIN,
+                     money_supply.getdouble() / (double) COIN,
+                     scaled_pool_magnitude.getdouble() / 100.0,
+                     scaled_network_magnitude.getdouble() / 100.0,
+                     pindex->nHeight,
+                     blocks,
+                     active_vote_weight_tally.getdouble() / (double) COIN,
+                     (active_vote_weight_tally / arith_uint256(blocks)).getdouble() / (double) COIN
+                     );
+
+            break;
+        }
+    }
+
```

Comment:
Maybe I'm nitpicking, but I'm thinking that when we reach this line we have benchmarked what we're interested in, so the code could be:

```C++
    g_timer.GetTimes("finished execution for poll id " + Txid().GetHex() + " title " + Title(), __func__);

    if (!blocks) {
        LogPrintf("WARNING: %s: No blocks tallied for poll txid %s, title \"%s\", so no active vote weight can be "
                  "calculated.",
                 __func__,
                 Txid().GetHex(),
                 Title()
                 );

        return std::nullopt;
    }

    arith_uint256 active_vote_weight = active_vote_weight_tally / arith_uint256(blocks);
    
    // Overflow protection. This should never happen.
   ....
```

That is, finish the timer, then figure out whether to bail or not.

Reply:
I don't like that, because you are leaving out the 256 bit division at the end... which is not trivial.

Reply:
Maybe it is not important, but that might be a couple of msec or more, even on slow machines.

Reply:
I am fine with just a single timer entry before the division by zero protection.

Reply:
2021-07-11T19:14:09Z timer GetActiveVoteWeight: finished execution for poll id before divide by zerod5f6339e8a61426e5ea79d688ff2142bd67128e5a7e1bfe520ab82ea7b080c31 title which cutting-edge consensus algorithm should gridcoin adopt?: elapsed time: 597 ms, time since last check: 597 ms.
2021-07-11T19:14:09Z timer GetActiveVoteWeight: finished execution for poll id d5f6339e8a61426e5ea79d688ff2142bd67128e5a7e1bfe520ab82ea7b080c31 title which cutting-edge consensus algorithm should gridcoin adopt?: elapsed time: 597 ms, time since last check: 0 ms.

It is showing 0 ms. We can go with a single entry before the division.

Reply:
Done.

Reply:
I think the important bit is being able to capture performance regressions or unusual behaviors. The division should be fairly constant while a chain goof will show up much more clearly.

Reply:
Agreed.